File: cpp/CurrentDateTimeTest.cpp
#include "current_date_time.pb.h"

#include <cstring>
#include <iostream>
#include <iomanip>
#include <fstream>

using namespace std;
using namespace protobuftest;

void buildMessage(CurrentDateTimeMessage* msg) {
    msg->set_title("A simple test");
    msg->set_counter(42);
    CurrentDateTimeMessage::Date* date = msg->mutable_date();
    date->set_day(11);
    date->set_month(9);
    date->set_year(2001);
    date->set_weekday("Monday");
    CurrentDateTimeMessage::Time* time = msg->mutable_time();
    time->set_hours(1);
    time->set_minutes(2);
    time->set_seconds(3);
}

bool storeMessage(const CurrentDateTimeMessage* msg, const char* fname) {
    ofstream serializingStream(fname, ofstream::out | ofstream::binary | ofstream::trunc);
    return msg->SerializeToOstream(&serializingStream);
}

bool loadMessage(CurrentDateTimeMessage* msg, const char* fname) {
    ifstream serializingStream(fname, ofstream::binary);
    return msg->ParseFromIstream(&serializingStream);
}

void printMessage(const CurrentDateTimeMessage* msg) {
    cout << "Received a message:" << endl;
    cout << "\tTitle: " << msg->title() << endl;
    cout << "\tCounter: " << msg->counter() << endl;
    if (msg->has_date()) {
        const CurrentDateTimeMessage::Date& date = msg->date();
        cout << "\t" << date.weekday() << ", " << date.year() << "-" <<
            setw(2) << setfill('0') << date.month() << "-" <<
            setw(2) << setfill('0') << date.day() << endl;
    } else {
        cout << "No date element found!" << endl;
    }
    if (msg->has_time()) {
        const CurrentDateTimeMessage::Time& time = msg->time();
        cout << "\t" << setw(2) << setfill('0') << time.hours() << ":" <<
            setw(2) << setfill('0') << time.minutes() << ":" <<
            setw(2) << setfill('0') << time.seconds() << endl;
    } else {
        cout << "No time element found!" << endl;
    }
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        cerr << "Please provide a filename!" << endl;
        exit(1);
    }
    const char* fname = argv[1];

    ifstream f(fname);
    if (!f.good()) {
        cout << "Creating current date/time entry" << endl;
        CurrentDateTimeMessage msg;
        buildMessage(&msg);
        storeMessage(&msg, fname);
    }
    CurrentDateTimeMessage msg2;
    if (loadMessage(&msg2, fname)) {
        printMessage(&msg2);
    } else {
        cerr << "Failed to load message!" << endl;
    }

    return 0;
}

File: java/CurrentDateTimeTest.java
import com.example.protobuftest.CurrentDateTime.CurrentDateTimeMessage;
import java.io.*;

class CurrentDateTimeTest {

    CurrentDateTimeMessage buildMessage() {
        CurrentDateTimeMessage.Builder msg = CurrentDateTimeMessage.newBuilder();
        msg.setTitle("A simple test #2");
        msg.setCounter(43);
        CurrentDateTimeMessage.Date.Builder date = CurrentDateTimeMessage.Date.newBuilder();
        date.setDay(12);
        date.setMonth(10);
        date.setYear(2002);
        date.setWeekday("Tuesday");
        msg.setDate(date);
        return msg.build();
    }

    boolean storeMessage(CurrentDateTimeMessage msg, String fname) {
        boolean success = false;
        FileOutputStream output = null;
        try {
            output = new FileOutputStream(fname);
            msg.writeTo(output);
            output.close();
            success = true;
        } catch (Exception e) {
            ;
        }
        return success;
    }

    CurrentDateTimeMessage loadMessage(String fname) {
        CurrentDateTimeMessage msg = null;
        FileInputStream input = null;
        try {
            input = new FileInputStream(fname);
            msg = CurrentDateTimeMessage.parseFrom(input);
            input.close();
        } catch (Exception e) {
            ;
        }
        return msg;
    }

   void printMessage(CurrentDateTimeMessage msg) {
        System.out.println("Received a message:");
        System.out.printf("\tTitle: %s\n", msg.getTitle());
        System.out.printf("\tCounter: %d\n", msg.getCounter());
        if (msg.hasDate()) {
            CurrentDateTimeMessage.Date date = msg.getDate();
            System.out.printf("\t%s, %d-%02d-%02d\n", date.getWeekday(), date.getYear(), date.getMonth(), date.getDay());
        } else {
            System.out.println("No date element found!");
        }
        if (msg.hasTime()) {
            CurrentDateTimeMessage.Time time = msg.getTime();
            System.out.printf("\t%02d:%02d:%02d\n", time.getHours(), time.getMinutes(), time.getSeconds());
        } else {
            System.out.println("No time element found!");
        }
   }

   public static final void main(String[] args) {
       if (args.length == 0) {
           System.err.println("Please provide a filename!");
           System.exit(1);
       }
       CurrentDateTimeTest instance = new CurrentDateTimeTest();
       String fname = args[0];
       if (!new File(fname).exists()) {
           System.out.println("Creating current date/time entry");
           CurrentDateTimeMessage msg = instance.buildMessage();
           instance.storeMessage(msg, fname);
       }
       CurrentDateTimeMessage msg2 = instance.loadMessage(fname);
       instance.printMessage(msg2);
   }
}

